code/modules/mob/living/silicon/robot/robot.dm

```diff
@@ -312,12 +314,32 @@
 			add_overlay("ov-opencover +c")
 		else
 			add_overlay("ov-opencover -c")
+
 	if(hat)
-		var/mutable_appearance/head_overlay = hat.build_worn_icon(default_layer = 20, default_icon_file = 'icons/mob/clothing/head/default.dmi')
-		head_overlay.pixel_z += hat_offset
-		add_overlay(head_overlay)
+		hat_overlay = hat.build_worn_icon(default_layer = 20, default_icon_file = 'icons/mob/clothing/head/default.dmi')
+		update_worn_icons()
+	else if(hat_overlay)
+		QDEL_NULL(hat_overlay)
+
 	update_appearance(UPDATE_OVERLAYS)
 
+/mob/living/silicon/robot/proc/update_worn_icons()
+	if(!hat_overlay)
+		return
+	cut_overlay(hat_overlay)
+
```

Comment:
```suggestion
	if(!islist(hat_offset))
		return
```
This will fix that tophat using infinity think you mentioned
Otherwise it'll runtime error from trying to list access a number

And if a model wants no hat sprites in the future, they could just do `hat_offset = null` (it would work the same as `hat_offset = INFINITY`)

Reply:
Shouldn't ever get a hat worn if this case happens, but added the islist check below just in case.